fix(column)!: ensure 'statuscolumn' works with virtual and wrapped lines

luukvbaal opened the pull request:
**fix(column)!: ensure 'statuscolumn' works with virtual and wrapped lines**

Problem:    The `'statuscolumn'` was not re-evaluated for wrapped lines, when preceded by virtual/filler lines. There was also no way to distinguish virtual and wrapped lines in the status column.

Solution:   Make sure to rebuild the statuscolumn, and replace variable `v:wrap` with `v:virtnum`. `v:virtnum` is negative when drawing virtual lines, zero when drawing the actual buffer line, and positive when drawing the wrapped part of a buffer line.

Also resolve a memory leak when returning from `win_line()` by `break`ing instead of `return`ing.

**fix(column): avoid drawing columns for virt_lines_leftcol**

Problem:    The default fold column, as well as the 'statuscolumn', were drawn unnecessarily/unexpectedly for virtual lines placed with `virt_lines_leftcol` set.

Solution:   Skip the column states if a virtual line with `virt_lines_leftcol` set will be drawn.

luukvbaal committed:
fix(column)!: ensure 'statuscolumn' works with virtual and wrapped lines

Problem:    The `'statuscolumn'` was not re-evaluated for wrapped lines,
            when preceded by virtual/filler lines. There was also no way
            to distinguish virtual and wrapped lines in the status column.
Solution:   Make sure to rebuild the statuscolumn, and replace variable
            `v:wrap` with `v:virtnum`. `v:virtnum` is negative when drawing
            virtual lines, zero when drawing the actual buffer line, and
            positive when drawing the wrapped part of a buffer line.

luukvbaal commented:
I guess this actually revealed an issue, only tested with diff-mode filler lines before.

luukvbaal commented:
> It's hard to tell from the test, but statuscolumn should not be rendered at all for virtual lines.

Hmm why? Isn't the point of `statuscolumn` to let the user decide how their "gutter" will look? Seems arbitrary to exclude virtual lines from this, e.g. with a column border:
![image](https://user-images.githubusercontent.com/31730729/212066938-9c51bf54-b781-4124-bf61-a0fdce0274f3.png)

I would be more in favor of introducing a `v:virtline` (or something) variable if it needs to be distinguishable from `v:wrap`.

lewis6991 commented:
Hmm ok maybe, but we need to make sure `virt_lines_leftcol` is properly handled.

luukvbaal commented:
Oof, seems to work actually.

lewis6991 commented:
> Oof, seems to work actually. However it is drawn over the statuscolumn, is that okay?

Not really.

>  Or should statuscolumn not be drawn for virt_lines_leftcol lines?

Yes

luukvbaal commented:
> > Oof, seems to work actually. However it is drawn over the statuscolumn, is that okay?
> 
> Not really.
> 
> > Or should statuscolumn not be drawn for virt_lines_leftcol lines?
> 
> Yes

The same is true for the default sign column BTW, should that be considered a bug? This is some `virt_lines_leftcol` lines without 'statuscolumn' set:
![image](https://user-images.githubusercontent.com/31730729/212345637-a6d0945b-803f-408d-95c6-65b35a6895e3.png)

luukvbaal commented:
> I would be more in favor of introducing a `v:virtline` (or something) variable if it needs to be distinguishable from `v:wrap`.

### Proposal

Rename `v:wrap` to `v:virtline`(?) and set it's value to:
* `-1` for filler_lines above the actual buffer line.
* `0` for the actual buffer line
* `1` for the wrapped part of lines

Alternatively we could set it to `row - startrow - filler_lines` to have it indicate the distance to the actual buffer line. Not sure that that would be useful but I guess it wouldn't hurt.

User can check that `v:fillnum == / < / > 0`. Seems better than adding a new variable.

zeertzjq commented on runtime/doc/eval.txt:
```diff
@@ -2295,13 +2295,15 @@ v:version	Vim version number: major version times 100 plus minor
+						*v:virtnum* *virtnum-variable*
+v:virtnum	Virtual line number for the 'statuscolumn' expressions.
+		A negative number indicates the number of virtual lines above
+		the start of the buffer line |v:lnum|, zero for the first text
```
Is this `v:lnum + 1` instead?

zeertzjq commented on runtime/doc/eval.txt:
```diff
@@ -2295,13 +2295,15 @@ v:version	Vim version number: major version times 100 plus minor
+v:virtnum	Virtual line number for the 'statuscolumn' expressions.
+		A negative number indicates the number of virtual lines above
+		the start of the buffer line |v:lnum|, zero for the first text
+		row of v:lnum, and postive for the wrapped part of a line.
```
Is this positive number always 1?

luukvbaal replied:
Currently yeah, don't really think it's worth re-evaluating statuscolumn for each line wrap.

luukvbaal commented:
Do we need `newst.txt` updates for breaking changes on unreleased features?

zeertzjq commented:
This change is probably to small to mention as these variables aren't mentioned either.

zeertzjq commented on src/nvim/drawline.c:
```diff
@@ -400,40 +400,45 @@ static int get_sign_attrs(buf_T *buf, linenr_T lnum, SignTextAttrs *sattrs, int
 ///
 /// @param[out] stcp  Status column attributes
 static void get_statuscol_str(win_T *wp, linenr_T lnum, int row, int startrow, int filler_lines,
                               int cul_attr, int sign_num_attr, int sign_cul_attr, char_u *extra,
```
It seems that `extra` is unused in this function and `get_statuscol_display_info`

luukvbaal replied:
Right, remnant from previous iteration of the original PR. Shouldn't this give a lint error?

luukvbaal committed:
fix(column): avoid drawing columns for virt_lines_leftcol

Problem:    The default fold column, as well as the 'statuscolumn', were
            drawn unnecessarily/unexpectedly for virtual lines placed
            with `virt_lines_leftcol` set.
Solution:   Skip the column states if a virtual line with
            `virt_lines_leftcol` set will be drawn.

luukvbaal commented:
https://github.com/neovim/neovim/pull/21768/commits/e627053622b21419e330cf80cf9b2ee3da82ff53 fixes this issue https://github.com/neovim/neovim/pull/21768#issuecomment-1381945101. It might be better discussed in a separate PR but it depends on this so I added it here.

lewis6991 reviewed (approved):
LGTM

zeertzjq commented on test/functional/ui/statuscolumn_spec.lua:
```diff
@@ -4,6 +4,7 @@ local clear = helpers.clear
 local command = helpers.command
 local eq = helpers.eq
 local eval = helpers.eval
+local exec = helpers.exec_lua
```
This looks a bit confusing. Why not name this variable `exec_lua`?

zeertzjq commented on test/functional/ui/statuscolumn_spec.lua:
```diff
@@ -183,7 +184,7 @@ describe('statuscolumn', function()
 
   it('works with wrapped lines, signs and folds', function()
-    command("set stc=%C%s%=%{v:wrap?'':v:lnum}│\\ ")
+    command("set stc=%C%s%=%{v:virtnum?'':v:lnum}│\\ ")
```
Since you change the command above to use long brackets, maybe change this as well:
```suggestion
    command([[set stc=%C%s%=%{v:virtnum?'':v:lnum}│\ ]])
```

zeertzjq commented on test/functional/ui/statuscolumn_spec.lua:
```diff
@@ -234,7 +235,7 @@ describe('statuscolumn', function()
     command('norm zf$')
     -- Check that alignment works properly with signs after %=
-    command("set stc=%C%=%{v:wrap?'':v:lnum}│%s\\ ")
+    command("set stc=%C%=%{v:virtnum?'':v:lnum}│%s\\ ")
```
And this
```suggestion
    command([[set stc=%C%=%{v:virtnum?'':v:lnum}│%s\ ]])
```

zeertzjq commented on test/functional/ui/statuscolumn_spec.lua:
```diff
@@ -385,25 +411,36 @@ describe('statuscolumn', function()
-    command('set stc=%C fdc=9 fillchars=foldsep:𒀀')
+  it('works with foldcolumn', function()
+    -- Fits maximum multibyte foldcolumn #21759
+    command('set stc=%C%=%l\\  fdc=9 fillchars=foldsep:𒀀')
```
And this
```suggestion
    command([[set stc=%C%=%l\  fdc=9 fillchars=foldsep:𒀀]])
```

phgz commented:
> Alternatively we could set it to row - startrow - filler_lines(v:virtnum) to have it indicate the distance to the actual buffer line. Not sure that that would be useful but I guess it wouldn't hurt.

This could be nice to have actually. I am trying to do a different action depending on the relative wrapped line position to the actual buffer line. Also, it would not be breaking change.

luukvbaal commented:
I would be okay with this. It would mean 'statuscolumn' is re-evaluated for each line in the window, currently it is only re-evaluated once for filler lines above the buffer line, once for the buffer line itself, and once for the wrapped lines below. So there would be a performance hit as well, but not compared to buffers where 'wrap' is disabled anyways. So not something we have to worry about I think.
